forte/processors/data_augment/algorithms/UDA.py

```diff
@@ -0,0 +1,192 @@
+# Copyright 2020 The Forte Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file implements the UDA(Unsupervised Data Augmentation)
+algorithm as described in
+Unsupervised Data Augmentation for Consistency Training
+(https://arxiv.org/abs/1904.12848)
+(https://github.com/google-research/uda)
+"""
+from typing import Optional, List, Callable, Tuple, Iterator
+from torch import Tensor
+from texar.torch.losses.info_loss import kl_divg_loss_with_logits
+from texar.torch.data import DataIterator
+from texar.torch.data.data.dataset_utils import Batch
+
+
+class UDAIterator:
+    r"""
+    This iterator wraps the Unsupervised Data Augmentation(UDA)
+    algorithm by calculating the unsupervised loss automatically
+    during each iteration. It takes both supervised and unsupervised
+    data iterator as input.
+
+    The unsupervised data should contain the original input and the
+    augmented input. The original and augmented inputs should be in
+    the same training example. The user should provide a function
+    unsup_forward_fn to calculate the output logits for original &
+    augmented inputs from the unsupervised batch.
+
+    During each iteration, the iterator will calculate the UDA loss
+    automatically with the given unsup_forward_fn. User can combine the
+    UDA loss with the supervised loss for model training.
+
+    It uses tricks such as prediction sharpening and confidence masking.
+    Please refer to the UDA paper for more details.
+    (https://arxiv.org/abs/1904.12848)
+
```

Comment:
Can you check if the generated docs are working well? I am not sure whether `-` will affect here.